torchio/metrics/map_metric.py

```diff
@@ -0,0 +1,47 @@
+from .base_metric import Metric
+from ..data.subject import Subject
+from ..torchio import DATA
+from typing import Sequence, Union
+from abc import ABC
+import numpy as np
+import torch
+import warnings
+
+average_func_dict = {"mean": torch.mean,
+                     "sum": torch.sum,
+                     }
+
+
+class MapMetric(Metric, ABC):
+
+    def __init__(self, mask_keys: Sequence = [], average_method: str = None):
+        super(MapMetric, self).__init__()
+        self.mask_keys = mask_keys
+        self.average_method = self._get_average_method(average_method)
+
+    @staticmethod
+    def _get_average_method(average_method: str):
+        if average_method is None:
+            return None
+        if average_method in average_func_dict.keys():
+            return average_func_dict[average_method]
+        else:
+            warnings.warn("Unfound specified averaging method. Averaging methods available are: {}\nUsing mean function")
+            return average_func_dict["mean"]
+
+    def _apply_masks_and_averaging(self, sample: Subject, metric_map: Union[torch.Tensor, np.ndarray]):
+        masked_map = metric_map
+        if self.mask_keys:
+            mask_keys_in_sample = sample.keys() & set(self.mask_keys)
+
+            if len(mask_keys_in_sample) == 0:
+                warnings.warn("None of the given masks {} found for the sample".format(self.mask_keys))
+            else:
+
+                for mask_key in mask_keys_in_sample:
+                    mask_data = sample[mask_key][DATA]
+                    masked_map = torch.mul(mask_data, metric_map)
+
+        if self.average_method is not None:
+            masked_map = self.average_method(masked_map)
+        return masked_map
```

Comment:
la gestion des mask n'est toujours pas bonne (je crois), car tu fais la moyenne en prennant en compte tous les zeros ..
toi tu fais :
metrics_result = torch.mean( torch.mul(mask_data, metric_map))
si metric_map = [2, 2, 2, 2] et que le mask=[0, 0, 1,1]
ca va faire mean([0,0,2,2]) = 1
moi je veux mean([2,2]) = 2

donc il faut faire 

metrics_result = torch.mean(metric_map[mask_data>0] ) 
et avec une list de mask
metrics_result = [ torch.mean(metric_masp[mm>0] for mm in mask_data_list] 

ensuite il faut construire un dictionaire avec les champs :  nom de la metric + '_' + mask_name

sinon tu peux garde the self.average_method, si tu change la boucle for 41 -> 43 par
masked_map = [ metric_map[sample[mask_key][DATA] ]  for mask_key in mask_keys_in_sample ]

il reste juste a construire le dic de resultat

c'est clair ?